<h1 align=center><font size = 5>Peer-graded Assignment: Build a Regression Model in Keras</font></h1>


Author: Konstantinos Kontogeorgopoulos

Final Assignment for Introduction to Deep Learning & Neural Networks with Keras, 
IBM Skills Network

In [38]:
#Install libraries if you do not already have them
#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
#!pip install keras==2.1.6

## Lets first Download and clean the dataset

In [39]:
# We import the pandas and numpy libraries
import pandas as pd
import numpy as np

We read the dataset provided by the assignment https://cocl.us/concrete_data  
I downloaded it into the same folder as this FinalAssignment.ipynb

In [40]:
#read .csv data file and put it into concrete_data pandas dataframe object
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
#lets see the first 5 rows of the csv
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

This dataset is a concrete samples dataset.  
I see that i have Cement, Blast Furnace Slag, Fly ash, Water, Superplasticizer, Coarse Aggrepate, Fine Aggregate, Age and Strength as the columns

I check how many data points i have

In [41]:
concrete_data.shape

(1030, 9)

And i see that i have 1030 samples with the 9 columns that i talked about above.

Let's check if the data needs cleaning.

In [42]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [43]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

There are no NULL values in our data.  
So our data is clean and we do not need to clean them.

## Now that i know my data is clean i proceed to split the data into predictors and target according to the assignment's instructions

The target is Strength and all other columns are predictors

In [44]:
#Pass into the variable concrete_data_columns the concrete_data columns
concrete_data_columns = concrete_data.columns
#Pass into predictors variable the predictors, everything  except strength
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]
#Pass into the target variable strength 
target = concrete_data['Strength']

And i doublecheck the predictors

In [45]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

And the target

In [46]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

## A: Build a Baseline Model

So lets import Keras using tensorflow 2.11

In [47]:
import tensorflow as tf 
# print(tf.version.VERSION)
# from tensorflow import keras

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

In [48]:
# Regression model
def regression_model():
    #I create the model
    model = tf.keras.Sequential()
    # In Dense method i put in the hidden layer 10 neurons with the RelU activation function
    model.add(tf.keras.layers.Dense(10, activation='relu', input_shape=(8,)))
    model.add(tf.keras.layers.Dense(1))
    
    # and add the adam optimizer and the mean squared error as the loss function 
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn.

In [49]:
#import train_test_split from sklearn
from sklearn.model_selection import train_test_split
#and split the data into a training and test sets with 70-30 % training - test
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=7)

2. Train the model on the training data using 50 epochs.

In [50]:
# I call the method to create the model and put it into model var
model = regression_model()
# and i use 50 epochs to train it
model.fit(X_train, y_train, epochs=50, verbose=1)

Epoch 1/50
23/23 [==============================] - 0s 773us/step - loss: 23546.3633
Epoch 2/50
23/23 [==============================] - 0s 818us/step - loss: 4381.2383
Epoch 3/50
23/23 [==============================] - 0s 909us/step - loss: 3397.0249
Epoch 4/50
23/23 [==============================] - 0s 818us/step - loss: 2864.5684
Epoch 5/50
23/23 [==============================] - 0s 773us/step - loss: 2551.1487
Epoch 6/50
23/23 [==============================] - 0s 773us/step - loss: 2285.6899
Epoch 7/50
23/23 [==============================] - 0s 818us/step - loss: 2079.0208
Epoch 8/50
23/23 [==============================] - 0s 864us/step - loss: 1891.8784
Epoch 9/50
23/23 [==============================] - 0s 773us/step - loss: 1739.9865
Epoch 10/50
23/23 [==============================] - 0s 773us/step - loss: 1604.5944
Epoch 11/50
23/23 [==============================] - 0s 818us/step - loss: 1481.6110
Epoch 12/50
23/23 [==============================] - 0s 773us/step - loss

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [51]:
#evaluate the model 
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 778us/step


188.70938110351562

In [52]:
# import mean_squared_error from sklearn.metrics
from sklearn.metrics import mean_squared_error
# compute the mean squared error between the predicted concrete stregth y_pred and the actual concrete strength y_test 
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

188.7094222236994 0.0


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.  
5. Report the mean and the standard deviation of the mean squared errors.

In [53]:
all_mean_squared_errors = 50
#creating empty array
mean_squared_errors = []
#for 50 times i do the same as i did before
for i in range(0, all_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=50, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " + str(all_mean_squared_errors) + " mean squared errors without normalized data. \n")
print("Mean: "+ str(mean))
print("Standard Deviation: "+ str(standard_deviation))

MSE 1: 104.62203216552734
10/10 [==============================] - 0s 667us/step
MSE 2: 124.27359008789062
10/10 [==============================] - 0s 667us/step
MSE 3: 110.09347534179688
10/10 [==============================] - 0s 667us/step
MSE 4: 124.68961334228516
10/10 [==============================] - 0s 778us/step
MSE 5: 122.65692901611328
10/10 [==============================] - 0s 778us/step
MSE 6: 107.96141815185547
10/10 [==============================] - 0s 667us/step
MSE 7: 135.6141815185547
10/10 [==============================] - 0s 667us/step
MSE 8: 110.59779357910156
10/10 [==============================] - 0s 778us/step
MSE 9: 132.67596435546875
10/10 [==============================] - 0s 667us/step
MSE 10: 111.27578735351562
10/10 [==============================] - 0s 778us/step
MSE 11: 103.08679962158203
10/10 [==============================] - 0s 667us/step
MSE 12: 100.3890151977539
10/10 [==============================] - 0s 667us/step
MSE 13: 115.1204833984375
1